File: languages/java/app/src/test/java/org/benchmarkplayground/operations/QuickSortTest.java
package org.benchmarkplayground.operations;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

public class QuickSortTest {
    @ParameterizedTest
    @MethodSource("provideStringListsForQuickSort")
    public void quickSortTest_StringLists(List<String> unsorted, List<String> expectedResult, String testMessage) {
        List<String> actualResult = QuickSort.quickSort(unsorted);
        assertArrayEquals(expectedResult.toArray(), actualResult.toArray(), testMessage);
    }

    @Test
    public void quickSortTest_IntegerLists() {
        List<Integer> unsorted = new ArrayList<>(List.of(4, -54, 40, 400, 2, -7, 0, 1, 4));
        List<Integer> expectedResult = new ArrayList<>(List.of(-54, -7, 0, 1, 2, 4, 4, 40, 400));
        List<Integer> actualResult = QuickSort.quickSort(unsorted);
        assertArrayEquals(expectedResult.toArray(), actualResult.toArray(), "QuickSort: Integer Test");

    }

    private static Stream<Arguments> provideStringListsForQuickSort() {
        return Stream.of(
            // Zero elements
            Arguments.of(new ArrayList<>(), new ArrayList<>(), "QuickSort: Zero Elements Test"),
            // One elements
            Arguments.of(
                new ArrayList<>(List.of("badger")),
                new ArrayList<>(List.of("badger")),
                "QuickSort: One Element Test"
            ),
            // Multiple odd elements
            Arguments.of(
                new ArrayList<>(List.of("zinc", "trap", "tree", "trait", "badger", "apple", "chemistry", "pig", "zoo")),
                new ArrayList<>(List.of("apple", "badger", "chemistry", "pig", "trait", "trap", "tree", "zinc", "zoo")),
                "QuickSort: Multiple Odd Elements Test"
            ),
            // Multiple even elements
            Arguments.of(
                new ArrayList<>(List.of("trap", "apple", "chemistry", "pig", "zoo", "tree", "trait", "badger", "zinc", "crack")),
                new ArrayList<>(List.of("apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo")),
                "QuickSort: Multiple Even Elements Test"
            )
        );
    }
}

File: languages/c++/test/quick_sort_test.cpp
#define DOCTEST_CONFIG_IMPLEMENT
#include <doctest/doctest.h>
#include <cstring> // for strcmp

#include "operations.h"

using namespace operations;

// Helper function to compare two char** arrays
bool arraysEqual(char** a, char** b, int size) {
    for (int i = 0; i < size; ++i) {
        if (strcmp(a[i], b[i]) != 0) {
            return false;
        }
    }
    return true;
}

TEST_CASE("QuickSort zero elements") {
    char *unsorted[] = { };
    char *expected_result[] = { };

    char **actual_result = quickSort(unsorted, 0);

    CHECK(arraysEqual(actual_result, expected_result, 0));
}

TEST_CASE("QuickSort single element") {
    char *unsorted[] = { "badger" };
    char *expected_result[] = { "badger" };

    char **actual_result = quickSort(unsorted, 1);

    CHECK(arraysEqual(actual_result, expected_result, 1));
}

TEST_CASE("QuickSort multiple odd elements") {
    char *unsorted[] = { "zinc", "trap", "tree", "trait", "badger", "apple", "chemistry", "pig", "zoo" };
    char *expected_result[] = { "apple", "badger", "chemistry", "pig", "trait", "trap", "tree", "zinc", "zoo" };

    char **actual_result = quickSort(unsorted, 9);

    CHECK(arraysEqual(actual_result, expected_result, 9));
}

TEST_CASE("QuickSort multiple even elements") {
    char *unsorted[] = { "trap", "apple", "chemistry", "pig", "zoo", "tree", "trait", "badger", "zinc", "crack" };
    char *expected_result[] = { "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };

    char **actual_result = quickSort(unsorted, 10);

    CHECK(arraysEqual(actual_result, expected_result, 10));
}

TEST_CASE("QuickSort already sorted elements") {
    char *unsorted[] = { "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };
    char *expected_result[] = { "apple", "badger", "chemistry", "crack", "pig", "trait", "trap", "tree", "zinc", "zoo" };

    char **actual_result = quickSort(unsorted, 10);

    CHECK(arraysEqual(actual_result, expected_result, 10));
}

int main(int argc, char* argv[]) {
    doctest::Context context;

    // defaults
    context.setOption("order-by", "file");            // sort the test cases by file and line

    // overrides
    context.setOption("no-breaks", true);               // don't break in the debugger when assertions fail

    int res = context.run();    // run queries, or run tests unless --no-run is specified

    if (context.shouldExit())   // important - query flags (and --exit) rely on the user doing this
        return res;             // propagate the result of the tests

    context.clearFilters();     // removes all filters added up to this point

    return res; // the result from doctest is propagated here as well
}
